codec/src/test/java/io/netty/handler/codec/FlowControlHandlerTest.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.handler.codec;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.hamcrest.collection.IsIterableContainingInOrder;
+import org.junit.Test;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class FlowControlHandlerTest {
+
+    /**
+     * This test is showing the default behavior. Auto reading is turned off but
+     * we still receive messages because ChannelHandler's are not subject to it.
+     */
+    @Test
+    public void testNettyAutoReadOff() {
+
+        final AtomicInteger counter = new AtomicInteger();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+            }
+        };
+
+        EmbeddedChannel channel = new EmbeddedChannel(handler);
+
+        channel.config().setAutoRead(false);
+        assertFalse(channel.config().isAutoRead());
+
+        channel.writeInbound("1", "2", "3");
+
+        // channelRead() was called three times even though auto reading was
+        // turned off
+        assertEquals(3, counter.get());
+    }
+
+    /**
+     * This is showing default behavior from a different but more common angle.
+     * The user receives one message and attempts to turn off auto reading
+     * inside the {@code channelRead()} call but continues receiving the
+     * remainder of the messages.
+     */
+    @Test
+    public void testNettyTurnAutoReadingOff() {
+
+        final AtomicInteger counter = new AtomicInteger();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+
+                // We're turning reading off but we'll continue receiving
```

Comment:
Again this is controlled via [RecvByteBufAllocator](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/RecvByteBufAllocator.java) for "regular channel types" (not channels like `EmbeddedChannel` :) ). The caveat is depending on where you are in the pipeline a single read from the socket may result in `[0,n)` messages in the pipeline.